CAP/gap/InstallAdds.gi

```diff
@@ -281,8 +271,74 @@ InstallGlobalFunction( CapInternalInstallAdd,
                     INSTALL_TODO_FOR_LOGICAL_THEOREMS( record.function_name, arg{ argument_list }, result, category );
                 fi;
                 
-                if (not is_derivation) and category!.add_primitive_output then
-                    add_function( category, result );
+                if (not is_derivation) then
+                    if category!.add_primitive_output then
+                        if record.return_type = "object" then
+                            AddObject( category, result );
+                        elif record.return_type = "morphism" then
+                            AddMorphism( category, result );
+                        elif record.return_type = "twocell" then
+                            AddTwoCell( category, result );
+                        elif record.return_type = "morphism_or_fail" then
+                            CAP_INTERNAL_ADD_MORPHISM_OR_FAIL( category, result );
+                        fi;
+                    elif category!.output_sanity_check_level > 0 then
+                        if record.return_type = "object" then
+                            if not IsCapCategoryObject( result ) then
+                                display_error_for_current_function( "the result does not lie in the filter IsCapCategoryObject." );
+                            fi;
+                            if not HasCapCategory( result ) then
```

Comment:
You could move out the `HasCapCategory` and `IsIdenticalObj(CapCategory...` `if`s out of the inner `if`s, if you do not want the stuff to be checked in a particular order. `HasCapCategory` can be applied to any object, since it is a filter.

Reply:
I think `result` could also be a boolean here, so not having a `CapCategory` would in general not be an error. I could move `IsIdenticalObj(CapCategory...` outside but then I would first have to check for `HasCapCategory` again, so I just kept this inside.

Reply:
Not really, because you already checked `HasCapCategory`, and if you did not have one, you got an error, so you would never get to the inside.

Reply:
As I understand it, you suggest the following code:
```
...
elif category!.output_sanity_check_level > 0 then
    if not HasCapCategory( result ) then
        CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result has no CAP category." );
    fi;
    if not IsIdenticalObj( CapCategory( result ), category ) then
        CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in this category." );
    fi;
    if record.return_type = "object" then
        if not IsCapCategoryObject( result ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in the filter IsCapCategoryObject." );
        fi;
        if not ObjectFilter( category )( result ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in the object filter of this category." );
        fi;
elif record.return_type = "morphism" then
...
```
This does not work, since `result` might be a boolean, for example for `IsWellDefined`. So calling `IsWellDefined` would always lead to the error `the result has no CAP category.`

If you suggest something else, I don't understand what exactly.

Reply:
I see the problem. You could tackle that problem by asking `IsAttributeStoringRep` at the beginning, at it is the minimal condition for all those tests anyway.

Reply:
If you suggest the following:
```
...
elif category!.output_sanity_check_level > 0 then
    if IsAttributeStoringRep( result ) then
        if not HasCapCategory( result ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result has no CAP category." );
        fi;
        if not IsIdenticalObj( CapCategory( result ), category ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in this category." );
        fi;
    fi;
    if record.return_type = "object" then
        if not IsCapCategoryObject( result ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in the filter IsCapCategoryObject." );
        fi;
        if not ObjectFilter( category )( result ) then
            CAP_INTERNAL_DISPLAY_ERROR_FOR_FUNCTION_OF_CATEGORY( record.function_name, category, "the result does not lie in the object filter of this category." );
        fi;
elif record.return_type = "morphism" then
...
```
This does not work if the `return_type` is `other_object`, because the other object does not have to fulfill `IsIdenticalObj( CapCategory( result ), category )`.

Reply:
You are right. Lets leave it how it is then.